mediachain/transactor/blockchain_follower.py

```diff
@@ -65,48 +68,82 @@ def cancel(self):
         self._cancelled = True
         self.block_ref_queue.put('__abort__')
         self.incoming_event_queue.put('__abort__')
+        # setting these events will cause the __abort__ messages to be
+        # pulled from the queues.
+        self.catchup_begin.set()
+        self.catchup_complete.set()
+
+    def _clear_queues(self):
+        with self.block_ref_queue.mutex:
+            self.block_ref_queue.queue.clear()
+        with self.block_replay_queue.mutex:
+            self.block_replay_queue.queue.clear()
+        with self.incoming_event_queue.mutex:
+            self.incoming_event_queue.queue.clear()
 
     def _perform_catchup(self):
         if not self.should_catchup:
             return
 
         while True:
+            # block until the event consumer thread tells us to start the
+            # blockchain catchup process
+            self.catchup_begin.wait()
+
             ref = self.block_ref_queue.get()
             if ref == '__abort__':
                 return
 
+            if ref_base58(ref) == self.last_known_block_ref:
+                print('hit last known block: {}'.format(
+                    self.last_known_block_ref
+                ))
+                self.catchup_complete.set()
+                continue
+
             block = self.cache.get(ref)
 
             if block is None:
+                # FIXME: we need to handle this better
```

Comment:
we could add a `fetch_if_missing` keyword arg that defaults to `False`, and make the consumer explicitly opt-in to remote fetching.